File: tests/nodejs/realworld_ssrf.js
/**
 * Real-world SSRF test cases for Kunlun-M scanner
 * Tests: http.get, http.request, https.get with user-controlled URL
 */

const express = require('express');
const http = require('http');
const https = require('https');
const { URL } = require('url');
const router = express.Router();

router.use(express.json());
router.use(express.urlencoded({ extended: true }));

/**
 * VULN: SSRF via http.get with req.query URL
 * Attacker can request internal services like http://169.254.169.254/latest/meta-data/
 * Example: /proxy?url=http://localhost:6379/ (Redis), http://169.254.169.254/metadata
 */
router.get('/proxy', (req, res) => {
    const targetUrl = req.query.url;
    http.get(targetUrl, (proxyRes) => {
        let body = '';
        proxyRes.on('data', (chunk) => body += chunk);
        proxyRes.on('end', () => res.send(body));
    }).on('error', (err) => {
        res.status(500).send('Proxy request failed');
    });
});

/**
 * VULN: SSRF via http.request with req.body url and method
 * Attacker controls both method and URL, allowing full internal scanning
 * Example POST: {"url": "http://127.0.0.1:22", "method": "GET"}
 */
router.post('/webhook-test', (req, res) => {
    const { url, method } = req.body;
    const parsedUrl = new URL(url);
    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || 80,
        path: parsedUrl.pathname,
        method: method || 'GET',
    };
    const proxyReq = http.request(options, (proxyRes) => {
        let body = '';
        proxyRes.on('data', (chunk) => body += chunk);
        proxyRes.on('end', () => res.json({ status: proxyRes.statusCode, body }));
    });
    proxyReq.on('error', () => res.status(500).send('Request failed'));
    proxyReq.end();
});

/**
 * VULN: SSRF via https.get with req.params URL
 * Attacker can probe internal HTTPS services
 * Example: /fetch/https://internal-admin:8443/admin
 */
router.get('/fetch/:targetUrl', (req, res) => {
    const encodedUrl = req.params.targetUrl;
    https.get(encodedUrl, (proxyRes) => {
        let body = '';
        proxyRes.on('data', (chunk) => body += chunk);
        proxyRes.on('end', () => res.send(body));
    }).on('error', () => res.status(502).send('Fetch failed'));
});

/**
 * SAFE: URL whitelist - only allow specific allowed domains
 * Prevents access to internal/localhost/metadata endpoints
 */
router.get('/safe-proxy', (req, res) => {
    const targetUrl = req.query.url;
    try {
        const parsed = new URL(targetUrl);
        const allowedHosts = ['api.github.com', 'api.example.com', 'api.publicdata.org'];
        // SAFE: Strict hostname whitelist prevents SSRF
        if (!allowedHosts.includes(parsed.hostname)) {
            return res.status(403).send('Domain not allowed');
        }
        // Also block private/internal IPs
        if (parsed.hostname === 'localhost' || /^10\./.test(parsed.hostname) ||
            /^172\.(1[6-9]|2\d|3[01])\./.test(parsed.hostname) ||
            /^192\.168\./.test(parsed.hostname)) {
            return res.status(403).send('Internal addresses not allowed');
        }
        http.get(targetUrl, (proxyRes) => {
            let body = '';
            proxyRes.on('data', (chunk) => body += chunk);
            proxyRes.on('end', () => res.send(body));
        }).on('error', () => res.status(502).send('Fetch failed'));
    } catch (e) {
        res.status(400).send('Invalid URL');
    }
});

/**
 * SAFE: Origin check - validate protocol and prevent private IP ranges
 */
router.post('/safe-webhook', (req, res) => {
    const { url } = req.body;
    try {
        const parsed = new URL(url);
        // SAFE: Only allow HTTPS to prevent internal network scanning
        if (parsed.protocol !== 'https:') {
            return res.status(400).send('Only HTTPS URLs allowed');
        }
        // Block internal/private IP addresses
        const ip = parsed.hostname;
        const privateRanges = ['localhost', '127.0.0.1', '0.0.0.0', '::1',
            '169.254.169.254', 'metadata.google.internal'];
        if (privateRanges.includes(ip) || ip.startsWith('10.') ||
            ip.startsWith('172.16.') || ip.startsWith('192.168.')) {
            return res.status(403).send('Private addresses blocked');
        }
        https.get(url, (proxyRes) => {
            let body = '';
            proxyRes.on('data', (chunk) => body += chunk);
            proxyRes.on('end', () => res.json({ status: proxyRes.statusCode, body }));
        }).on('error', () => res.status(502).send('Request failed'));
    } catch (e) {
        res.status(400).send('Invalid URL');
    }
});

module.exports = router;

File: tests/python/realworld_ssrf.py
"""
Real-world SSRF (Server-Side Request Forgery) test cases for Kunlun-M scanner.
Simulates a Flask application with URL fetching / webhook proxy endpoints.
"""

import re
import requests
from flask import Blueprint, request, jsonify

proxy_bp = Blueprint('proxy', __name__)

# Whitelist of allowed internal services
ALLOWED_INTERNAL_HOSTS = [
    'api.internal.local',
    'metrics.internal.local',
    'health.internal.local',
]


# ---------------------------------------------------------------------------
# VULN 1: SSRF via requests.get with user-controlled URL
# ---------------------------------------------------------------------------
@proxy_bp.route('/api/proxy/fetch', methods=['GET'])
def fetch_url():
    """URL fetcher / proxy endpoint."""
    url = request.args.get('url', '')
    # VULN: User-controlled URL passed directly to requests.get
    # Attacker can supply: url="http://169.254.169.254/latest/meta-data/" (AWS metadata)
    # or url="http://localhost:6379/" (Redis) or url="file:///etc/passwd"
    resp = requests.get(url, timeout=10)
    return jsonify({'status_code': resp.status_code, 'body': resp.text[:2000]})


# ---------------------------------------------------------------------------
# VULN 2: SSRF via requests.post with user-controlled URL in webhook
# ---------------------------------------------------------------------------
@proxy_bp.route('/api/webhooks/dispatch', methods=['POST'])
def dispatch_webhook():
    """Dispatch a webhook to a user-specified endpoint."""
    target_url = request.form.get('callback_url', '')
    payload = request.form.get('payload', '')
    headers = {'Content-Type': 'application/json'}
    # VULN: User-controlled callback URL with no validation
    # Attacker can supply: callback_url="http://169.254.169.254/latest/meta-data/iam/"
    resp = requests.post(target_url, data=payload, headers=headers, timeout=10)
    return jsonify({'status_code': resp.status_code})


# ---------------------------------------------------------------------------
# VULN 3: SSRF via requests.get with partially constructed URL
# ---------------------------------------------------------------------------
@proxy_bp.route('/api/proxy/screenshot', methods=['GET'])
def screenshot_url():
    """Take a screenshot of a webpage (simulated)."""
    domain = request.args.get('domain', '')
    path = request.args.get('path', '/')
    # VULN: User controls domain, enabling internal network scanning
    # Attacker can supply: domain="169.254.169.254", path="/latest/meta-data/"
    url = f'http://{domain}{path}'
    resp = requests.get(url, timeout=15)
    return jsonify({'status_code': resp.status_code, 'body_length': len(resp.text)})


# ---------------------------------------------------------------------------
# SAFE 1: URL whitelist check before making request
# ---------------------------------------------------------------------------
@proxy_bp.route('/api/internal/service', methods=['GET'])
def query_internal_service():
    """Query an internal service by name."""
    service_name = request.args.get('service', '')
    # SAFE: Only allowed internal hostnames are accepted
    if service_name not in ALLOWED_INTERNAL_HOSTS:
        return jsonify({'error': 'Service not allowed'}), 403
    url = f'http://{service_name}:8080/health'
    resp = requests.get(url, timeout=5)
    return jsonify({'status': resp.json()})


# ---------------------------------------------------------------------------
# SAFE 2: Origin/protocol validation with regex
# ---------------------------------------------------------------------------
@proxy_bp.route('/api/proxy/external', methods=['POST'])
def fetch_external():
    """Fetch an external resource with strict origin validation."""
    url = request.form.get('url', '')
    # SAFE: Validate that URL uses HTTPS and points to a public domain
    if not re.match(r'^https://([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}/', url):
        return jsonify({'error': 'Invalid URL: must be HTTPS with a public domain'}), 400
    parsed = requests.utils.urlparse(url)
    # Block private/internal IP ranges
    hostname = parsed.hostname
    if re.match(r'^(10\.|172\.(1[6-9]|2\d|3[01])\.|192\.168\.|127\.|169\.254\.|0\.|localhost)', hostname):
        return jsonify({'error': 'Private/internal addresses are not allowed'}), 403
    resp = requests.get(url, timeout=10)
    return jsonify({'status_code': resp.status_code})
